### PR TITLE
[GSoC2024] fix warped image loading

### PR DESCRIPTION
**Pull request by tahamukhtar20**
### Motivation and context

Fixes: #7394 

This PR fixes the issue stated in the above-mentioned issue, the issue was caused when a 90 deg-rotated video was added with "Prefer Zip Chunks" disabled, and the image loaded was warped, which is now fixed using the static cropImage function found in this [commit](https://github.com/opencv/cvat/commit/844a72b82c241c2f612debc113de54b2f3612e7e#diff-2ca4b22442f076f5f5c3086ee757649fbaa903a75baca70991e419e2bd1d822fR281) as this was the commit where this issue initially started

### How has this been tested?

Added various images to test the bug fix, this did not require any new test cases.

### Checklist

- [x] I submit my changes into the `develop` branch
- [x] I have linked related issues (see [GitHub docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/opencv/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.

### CODE REVIEW

**Comment by tahamukhtar20** [comment]
I didn't post a solution before for review, it skipped my mind that I had to do that, please let me know if there are any issues with this, and I'll reimplement that.
I've already implemented this solution in the PR attached above, here is how I solved it currently, I noticed that this was working fine in v-2.5.1 and not in v-2.6.2 so I checked out the commits in between both versions and identified that the culprit was this piece of cvat-data code change in this [commit](https://github.com/opencv/cvat/commit/844a72b82c241c2f612debc113de54b2f3612e7e#diff-2ca4b22442f076f5f5c3086ee757649fbaa903a75baca70991e419e2bd1d822fR281). Now what I did was identify what was missing here, I noticed that imageBitmap was now being used across the entire application but something was missing that was in the previous PR that wasn't in this. which was that the crop image function's code was completely missing and that was why the issue was happening in the first place. So I added that in the code base in a way that nothing else was affected and tested it out in various environments making sure it worked for all types of images

**Comment by tahamukhtar20** [comment]
@bsekachev Sorry for pinging you like this, I think you should also have a look at this Pull Request as the piece of code that I changed here was directly changed by you a few months back.

File: cvat-data/src/ts/cvat-data.ts [hunk 1]
```diff
@@ -239,8 +272,15 @@ export class FrameDecoder {
...
                     const width = Math.round(this.renderWidth / scaleFactor);
 
-                    const array = new Uint8ClampedArray(e.data.buf.slice(0, width * height * 4));
-                    createImageBitmap(new ImageData(array, width)).then((bitmap) => {
+                    createImageBitmap(FrameDecoder.cropImage(
+                        e.data.buf,
+                        e.data.width,
+                        e.data.height,
+                        0,
+                        0,
```
Comment (bsekachev) [inline cvat-data/src/ts/cvat-data.ts:280]: - We do not need constant parameters.
- Function needs to be rewritten based on this constant values (code may be reduced). Is not it?

Comment (tahamukhtar20) [reply cvat-data/src/ts/cvat-data.ts:280]: Thanks for the response, I'll look into it.

Comment (tahamukhtar20) [reply cvat-data/src/ts/cvat-data.ts]: @bsekachev please have a look now.

**Comment by nmanovic** [comment]
@SpecLad , could you please help with the review?

File: cvat-data/src/ts/cvat-data.ts [hunk 2]
```diff
@@ -200,6 +200,37 @@ export class FrameDecoder {
...
+        const bufferSize = width * height * 4;
+        const buffer = new ArrayBuffer(bufferSize);
+        const rgbaInt32 = new Uint32Array(buffer);
+        const rgbaInt8Clamped = new Uint8ClampedArray(buffer);
+
+        if (imageWidth === width) {
+            return new ImageData(new Uint8ClampedArray(imageBuffer, 0, bufferSize), width, height);
+        }
+
+        let writeIdx = 0;
```
Comment (bsekachev) [inline cvat-data/src/ts/cvat-data.ts:224]:
```suggestion
        if (imageWidth === width) {
            return new ImageData(new Uint8ClampedArray(imageBuffer, 0, bufferSize), width, height);
        }

        const buffer = new ArrayBuffer(bufferSize);
        const rgbaInt32 = new Uint32Array(buffer);
        const rgbaInt8Clamped = new Uint8ClampedArray(buffer);
        let writeIdx = 0;
```

Comment (bsekachev) [reply cvat-data/src/ts/cvat-data.ts:224]: Perhaps we only need to create buffer when we are going to use it

**Comment by bsekachev** [comment]
@tahamukhtar20 

Could you please check changelog entry?

``scriv create --edit``

**Comment by tahamukhtar20** [comment]
@bsekachev added the changelog entry.

File: changelog.d/20240326_195957_tahamukhtar20.md [hunk 3]
```diff
@@ -0,0 +1,4 @@
+### Fixed
+
+- 90 deg-rotated video was added with "Prefer Zip Chunks" disabled was warped, fixed using the static cropImage function.
```
Comment (bsekachev) [inline changelog.d/20240326_195957_tahamukhtar20.md:3]:
```suggestion
- 90 deg-rotated video was added with "Prefer Zip Chunks" disabled 
was warped, fixed using the static cropImage function.
```

**Review by bsekachev — APPROVED** [review]

**Comment by bsekachev** [comment]
Great, thank you for the contribution!

**Pull request merged**